### PR TITLE
Bootguard 1.0 support

### PR DESCRIPTION
**Pull request by zaolin**
* Provisioning of bootguard 1.0
* Merge of bg & cbnt provisioning tooling

### CODE REVIEW

File: pkg/provisioning/bootguard/bootguard.go [hunk 1]
```diff
@@ -0,0 +1,772 @@
...
+	var b BootGuard
+	if bpm == nil {
+		return nil, fmt.Errorf("manifest is nil")
+	}
+	var err error
+	b.Version, err = bgheader.DetectBGV(bpm)
+	if err != nil {
+		return nil, err
+	}
+	bpm.Seek(0, 0)
```
Comment (ChriMarMe) [inline pkg/provisioning/bootguard/bootguard.go:108]: nit

File: pkg/provisioning/bootguard/bootguard.go [hunk 2]
```diff
@@ -0,0 +1,772 @@
...
+	var b BootGuard
+	if km == nil {
+		return nil, fmt.Errorf("manifest is nil")
+	}
+	var err error
+	b.Version, err = bgheader.DetectBGV(km)
+	if err != nil {
+		return nil, err
+	}
+	km.Seek(0, 0)
```
Comment (ChriMarMe) [inline pkg/provisioning/bootguard/bootguard.go:138]: nit

File: pkg/provisioning/bootguard/bootguard.go [hunk 3]
```diff
@@ -0,0 +1,772 @@
...
+	var b BootGuard
+	if bpm != nil && km != nil {
+		return nil, fmt.Errorf("either both or one manifest is nil")
+	}
+	var err error
+	bpmV, err := bgheader.DetectBGV(bpm)
+	if err != nil {
+		return nil, err
+	}
+	bpm.Seek(0, 0)
```
Comment (ChriMarMe) [inline pkg/provisioning/bootguard/bootguard.go:168]: nit

File: pkg/provisioning/bootguard/bootguard.go [hunk 4]
```diff
@@ -0,0 +1,772 @@
...
+	kmV, err := bgheader.DetectBGV(km)
+	if err != nil {
+		return nil, err
+	}
+	km.Seek(0, 0)
```
Comment (ChriMarMe) [inline pkg/provisioning/bootguard/bootguard.go:173]: nit

File: pkg/provisioning/bootguard/bootguard.go [hunk 5]
```diff
@@ -0,0 +1,772 @@
...
+func (b *BootGuard) SignKM(signAlgo string, privkey crypto.PrivateKey) ([]byte, error) {
+	buf := new(bytes.Buffer)
+	switch b.Version {
+	case bgheader.Version10:
+		signAlgo, err := bg.GetAlgFromString(signAlgo)
+		if err != nil {
+			return nil, err
+		}
+		b.VData.BGkm.RehashRecursive()
+		b.VData.BGkm.WriteTo(buf)
```
Comment (ChriMarMe) [inline pkg/provisioning/bootguard/bootguard.go:426]: nit

File: pkg/provisioning/bootguard/bootguard.go [hunk 6]
```diff
@@ -0,0 +1,772 @@
...
+		if err := b.VData.BGkm.SetSignature(signAlgo, privkey.(crypto.Signer), unsignedKM); err != nil {
+			return nil, err
+		}
+	case bgheader.Version20:
+		signAlgo, err := cbnt.GetAlgFromString(signAlgo)
+		if err != nil {
+			return nil, err
+		}
+		b.VData.CBNTkm.RehashRecursive()
+		b.VData.CBNTkm.WriteTo(buf)
```
Comment (ChriMarMe) [inline pkg/provisioning/bootguard/bootguard.go:437]: nit

**Review by ChriMarMe — APPROVED** [review]
just nits. error checking on all.

**Review by ChriMarMe — APPROVED** [review]

**Review by walterchris — APPROVED** [review]

File: cmd/bg-prov/README.md [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+3. Show details of ACM
+```bash
+./bg-prov show-acm ./ACM/acm_signed.bin
+```
+
+4. Show all 
+```bash
+./bg-prov show-all ./firmware.rom
+```
```
Comment (flanfly) [inline cmd/bg-prov/README.md:177]: We need some kind of acknowledgment that the development was partially funded by NLnet.

**Review by flanfly — COMMENTED** [review]
Well it's mostly renaming stuff.

**Review by xaionaro — APPROVED** [review]

**Comment by ansiwen** [comment]
I believe something important got lost here:
`bg-prov: error: no Run() method found in hierarchy of bpm-gen-v-1 <bpm> <bios>`

Also see https://github.com/9elements/converged-security-suite/issues/367

**Pull request merged**